File: 5.js
// Dynamic Programming
// TODO: find a smaller O

var longestPalindrome = function (s) {
  map = new Map();
  var x = _longestPalindrome(s, 0, s.length - 1);
  debugger;
  if (x[1] < x[0]) {
    return "";
  } else {
    return s.substring(x[0], x[1] + 1);
  }
};
var map = new Map();
function _longestPalindrome (s, start, end) {
  if (start > end) {
    return [0, 0];
  }
  let stringLength = end - start + 1;
  if (stringLength <= 1) {
    return [start, end];
  }
  if (map.get(`${start}-${end}`)) {
    return map.get(`${start}-${end}`);
  }

  let answer1 = getSingleLongestPalindrome(s, start, end - 1);
  let answer2 = getSingleLongestPalindrome(s, start + 1, end);
  let answer3 = getSingleLongestPalindrome(s, start, end);

  let max = findMax(answer1, answer2, answer3);
  // let max = answer3;
  let maxSize = max[1] - max[0];
  if (maxSize >= stringLength - 1) {
    map[`${start}-${end}`] = max;
    return max;
  }

  let left = _longestPalindrome(s, start, stringLength % 2 == 0 ? end - 2 : (end - 2 + 1));
  let right = _longestPalindrome(s, start + 2, end);
  max = findMax(max, left, right);
  // let maxSize = max[1] - max[0];
  map.set(`${start}-${end}`, max);
  return max;
}

function getSingleLongestPalindrome(s, originLeft, originRight) {
  var stringLength = originRight - originLeft + 1;
  var left = stringLength % 2 == 0 ? Math.floor(stringLength / 2) - 1 : Math.floor(stringLength / 2);
  left = originLeft + left;
  var right = stringLength % 2 == 0 ? Math.floor(stringLength / 2) : Math.floor(stringLength / 2) + 1;
  right = originRight + 1 - Math.floor(right);

  while (left >= originLeft && right <= originRight) {
    if (s[left] != s[right]) {
      let result = [left + 1, right - 1];
      return result;
    }
    left--;
    right++;
  }
  let result = [originLeft, originRight];
  return result;
}

function findMax(a, b, c)
{
  let sizeA = a[1] - a[0];
  let sizeB = b[1] - b[0];
  let sizeC = c[1] - c[0];
  return ((sizeA > sizeB ? sizeA : sizeB) > sizeC) ? (sizeA > sizeB ? a : b) : c;
}

let a = "ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggg";
console.log(longestPalindrome(a));

File: 5.cpp
#include <iostream>
#include <string>
using namespace std;

class Solution {
public:
    string longestPalindrome(string s) {
      int stringLength = s.size();
      if (stringLength == 0 || stringLength == 1) {
        return s;
      }
      if (stringLength == 2) {
        if (s[0] == s[1])
          return s;
        else
          return s[0] + "";
      }
      string answer1 = getSingleLongestPalindrome(s.substr(0, stringLength - 1));
      string answer2 = getSingleLongestPalindrome(s.substr(1));
      string answer3 = getSingleLongestPalindrome(s);

      string max = findMax(answer1, answer2, answer3);
      if (max.size() > (stringLength % 2 == 0 ? stringLength / 2 : stringLength / 2 + 1))
      {
        return max;
      } else {
        string left = longestPalindrome(s.substr(0, stringLength % 2 == 0 ? stringLength - 2 : (stringLength - 2 + 1)));
        string right = longestPalindrome(s.substr(2));
        return findMax(max, left, right);
      }
    }
    string getSingleLongestPalindrome(string s) {
      int stringLength = s.size();
      int left = stringLength / 2 - 1;
      int right = stringLength % 2 == 0 ? stringLength / 2 : stringLength / 2 + 1;

      while (left >= 0 && right < stringLength) {
        if (s[left] != s[right]) {
          return s.substr(left + 1, right - left - 1);
        }
        left--;
        right++;
      }
      return s;
    }
    string findMax(string a, string b, string c)
    {
      int sizeA = a.size();
      int sizeB = b.size();
      int sizeC = c.size();
      return ((sizeA > sizeB ? sizeA : sizeB) > sizeC) ? (sizeA > sizeB ? a : b) : c;
    }
};

int main()
{
  Solution *s = new Solution();
  cout << s->longestPalindrome("civilwartestingwhetherthatnaptionoranynartionsoconceivedandsodedicatedcanlongendureWeareqmetonagreatbattlefiemldoftzhatwarWehavecometodedicpateaportionofthatfieldasafinalrestingplaceforthosewhoheregavetheirlivesthatthatnationmightliveItisaltogetherfangandproperthatweshoulddothisButinalargersensewecannotdedicatewecannotconsecratewecannothallowthisgroundThebravelmenlivinganddeadwhostruggledherehaveconsecrateditfaraboveourpoorponwertoaddordetractTgheworldadswfilllittlenotlenorlongrememberwhatwesayherebutitcanneverforgetwhattheydidhereItisforusthelivingrathertobededicatedheretotheulnfinishedworkwhichtheywhofoughtherehavethusfarsonoblyadvancedItisratherforustobeherededicatedtothegreattdafskremainingbeforeusthatfromthesehonoreddeadwetakeincreaseddevotiontothatcauseforwhichtheygavethelastpfullmeasureofdevotionthatweherehighlyresolvethatthesedeadshallnothavediedinvainthatthisnationunsderGodshallhaveanewbirthoffreedomandthatgovernmentofthepeoplebythepeopleforthepeopleshallnotperishfromtheearth") << endl;
  return 0;
}